### ISSUE TITLE
Unicorn 2 fails to emulate multiple M68K instructions

### ISSUE THREAD

**Issue by zznop** [unicorn-engine/unicorn#1502]
Hit CPU unhandled exception with these M68K instructions (using 2.0.0-rc4):

```
0641 0100                                         addi.w  #$100,d1
```

```
46FC 2700                                         move    #$2700,sr
```

```
51C8 FFF2                                         dbf     d0,loc_220
```

```
0A47 0000                                         eori.w  #0,d7
```

**Comment by wtdcode** [comment]
Thanks for your report and I would have a look ASAP.

**Comment by zznop** [comment]
Did some digging into the C source code. I found that in `qemu/target/m68k/cpu.c` that there are different instructions enabled for different M68K CPU models. By setting `uc->cpu_model` to `0`, it sets the CPU to M68000 and it looks like it emulates most of the instructions mentioned above correctly. The only issue is with the `move` instruction into `sr`. I haven't traced down the code responsible for this yet, but I plan to dig deeper when I have more time. For now, here's the patch I applied to force it to set the CPU to M68000:

```
diff --git a/qemu/target/m68k/unicorn.c b/qemu/target/m68k/unicorn.c
index 3930bc56..8709ba5c 100644
--- a/qemu/target/m68k/unicorn.c
+++ b/qemu/target/m68k/unicorn.c
@@ -151,6 +151,7 @@ static int m68k_cpus_init(struct uc_struct *uc, const char *cpu_model)
 {
     M68kCPU *cpu;
 
+    uc->cpu_model = 0;
     cpu = cpu_m68k_init(uc);
     if (cpu == NULL) {
         return -1;
```

Obviously, this isn't a solution. I think we want the go bindings (and all other bindings) to expose the `cpu_model` in the unicorn interface/struct so we can set the processor model

**Comment by aquynh** [comment]
obviously, you are one of few people who seriously play with m68k.

do you have a similar issue with Unicorn 1?

**Comment by wtdcode** [comment]
> @wtdcode Did some digging into the C source code. I found that in `qemu/target/m68k/cpu.c` that there are different instructions enabled for different M68K CPU models. By setting `uc->cpu_model` to `0`, it sets the CPU to M68000 and it looks like it emulates most of the instructions mentioned above correctly. The only issue is with the `move` instruction into `sr`. I haven't traced down the code responsible for this yet, but I plan to dig deeper when I have more time. For now, here's the patch I applied to force it to set the CPU to M68000:
> 
> ```
> diff --git a/qemu/target/m68k/unicorn.c b/qemu/target/m68k/unicorn.c
> index 3930bc56..8709ba5c 100644
> --- a/qemu/target/m68k/unicorn.c
> +++ b/qemu/target/m68k/unicorn.c
> @@ -151,6 +151,7 @@ static int m68k_cpus_init(struct uc_struct *uc, const char *cpu_model)
>  {
>      M68kCPU *cpu;
>  
> +    uc->cpu_model = 0;
>      cpu = cpu_m68k_init(uc);
>      if (cpu == NULL) {
>          return -1;
> ```
> 
> Obviously, this isn't a solution. I think we want the go bindings (and all other bindings) to expose the `cpu_model` in the unicorn interface/struct so we can set the processor model

The API to set cpu model is here https://github.com/unicorn-engine/unicorn/blob/10d88e89a5648ee3ea90e20565e7cd6207484b1c/include/unicorn/unicorn.h#L598

**Comment by zznop** [comment]
> obviously, you are one of few people who seriously play with m68k.
> 
> 
> 
> do you have a similar issue with Unicorn 1?

Yes, I had a few issues open for unsupported M68K instructions with unicorn 1 in the past. The big difference is in UC1 the version of qemu unicorn was based off of only had support for 3 processors: M5206, M5208, and CFV4E. Now unicorn 2 looks like it should support all the common M68K variants, including actual M68000. 👏

**Comment by aquynh** [comment]
If M68000 is the most popular model, we can set it as the default model.

**Comment by zznop** [comment]
I'm not sure that M68000 is the most popular M68K variant anymore, but it makes sense that it should be the default (to me anyways). What I plan to submit a PR for, once I find some free time, is for the `uc_ctl_set_cpu_model`  API to be exposed to the Go bindings. In Go, there isn't a way to use C macros, which is how `uc_ctl_set_cpu_model` is implemented currently. So I'll need to write a little wrapper function in C that can be exported, so I can then write a `SetCPUModel` Go binding.

**Comment by wtdcode** [comment]
> I'm not sure that M68000 is the most popular M68K variant anymore, but it makes sense that it should be the default (to me anyways). What I plan to submit a PR for, once I find some free time, is for the `uc_ctl_set_cpu_model` API to be exposed to the Go bindings. In Go, there isn't a way to use C macros, which is how `uc_ctl_set_cpu_model` is implemented currently. So I'll need to write a little wrapper function in C that can be exported, so I can then write a `SetCPUModel` Go binding.

Looking forward to your PR. ;)

**Comment by zznop** [comment]
> > I'm not sure that M68000 is the most popular M68K variant anymore, but it makes sense that it should be the default (to me anyways). What I plan to submit a PR for, once I find some free time, is for the `uc_ctl_set_cpu_model` API to be exposed to the Go bindings. In Go, there isn't a way to use C macros, which is how `uc_ctl_set_cpu_model` is implemented currently. So I'll need to write a little wrapper function in C that can be exported, so I can then write a `SetCPUModel` Go binding.
> 
> Looking forward to your PR. ;)

It's up! I still need to look into what's up with the SR register though and why I can't emulate `move    <literal>,sr` I have a feeling #1501 is related

**Comment by zznop** [comment]
Finally had time to look into why emulation of the following instruction fails:

```
46FC 2700                                         move    #$2700,sr
```

There is no support for the `46fc` opcode in `translate.c`. There is support for `46c0`, which is move a register value into `sr`, but not `46fc`, move a literal/constant into `sr`. How would you like me to proceed? I can open a PR to add support for this opcode in unicorn. Or, would you rather see support added for this instruction in the `qemu` project, first?

**Comment by wtdcode** [comment]
> Finally had time to look into why emulation of the following instruction fails:
> 
> ```
> 46FC 2700                                         move    #$2700,sr
> ```
> 
> There is no support for the `46fc` opcode in `translate.c`. There is support for `46c0`, which is move a register value into `sr`, but not `46fc`, move a literal/constant into `sr`. How would you like me to proceed? I can open a PR to add support for this opcode in unicorn. Or, would you rather see support added for this instruction in the QEMU project, first?

Thanks for your effort. Sorry for my late reply.

It's not due to the lack of decoding but wrong CPU mode. The reason is that by default M68K emulation by Unicorn is not in user mode and thus the instruction is not allowed to execute. I inspect qemu code and you may clear the corresponding MSR register flags to enter usermode.

**Comment by wtdcode** [comment]
> > Finally had time to look into why emulation of the following instruction fails:
> > ```
> > 46FC 2700                                         move    #$2700,sr
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There is no support for the `46fc` opcode in `translate.c`. There is support for `46c0`, which is move a register value into `sr`, but not `46fc`, move a literal/constant into `sr`. How would you like me to proceed? I can open a PR to add support for this opcode in unicorn. Or, would you rather see support added for this instruction in the QEMU project, first?
> 
> Thanks for your effort. Sorry for my late reply.
> 
> It's not due to the lack of decoding but wrong CPU mode. The reason is that by default M68K emulation by Unicorn is not in user mode and thus the instruction is not allowed to execute. I inspect qemu code and you may clear the corresponding MSR register flags to enter usermode.

Sorry I'm wrong. By default Unicorn is in usermode and you need to switch to kernel mode to execute the instruction.

**Comment by wtdcode** [comment]
Thanks for your efforts which help me solve it easily. Fixed in 8a0ca8715e920d728d872584e5034583753f0d8b (unfortunately, your previous PR is not correct).

**Comment by zznop** [comment]
I think that's a separate issue. If you look at where are the instructions are registered, there's only support for `mov <reg>, sr` (`46c0`) - here: https://github.com/unicorn-engine/unicorn/blob/master/qemu/target-m68k/translate.c#L2942

I see what you're talking about the `IS_USER` macro/check. But that applies to the `move_to_sr` function, which is only registered for the `46c0` opcode. However, `46fc` (`mov <literal>, sr`) is not a supported opcode from what I can tell.
